File: lib/models/colis.dart
class Colis {
  final int id;
  final int clientId;
  final int? chauffeurId;
  final int stationDepartId;
  final int stationArriveeId;
  final String description;
  final double poids;
  final double? prix;
  final String statut;
  final DateTime? dateEnvoi;
  final DateTime? dateLivraison;
  final String? codeRetrait;
  final String? nomDestinataire;
  final String? numeroDestinataire;

  Colis({
    required this.id,
    required this.clientId,
    this.chauffeurId,
    required this.stationDepartId,
    required this.stationArriveeId,
    required this.description,
    required this.poids,
    this.prix,
    required this.statut,
    this.dateEnvoi,
    this.dateLivraison,
    this.codeRetrait,
    this.nomDestinataire,
    this.numeroDestinataire,
  });

  factory Colis.fromJson(Map<String, dynamic> json) {
    return Colis(
      id: json['id'],
      clientId: json['client_id'],
      chauffeurId: json['chauffeur_id'],
      stationDepartId: json['station_depart_id'],
      stationArriveeId: json['station_arrivee_id'],
      description: json['description'],
      poids: (json['poids'] as num).toDouble(),
      prix: json['prix'] != null ? (json['prix'] as num).toDouble() : null,
      statut: json['statut'],
      dateEnvoi: json['date_envoi'] != null ? DateTime.parse(json['date_envoi']) : null,
      dateLivraison: json['date_livraison'] != null ? DateTime.parse(json['date_livraison']) : null,
      codeRetrait: json['code_retrait'],
      nomDestinataire: json['nom_destinataire'],
      numeroDestinataire: json['numero_destinataire'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'client_id': clientId,
      'chauffeur_id': chauffeurId,
      'station_depart_id': stationDepartId,
      'station_arrivee_id': stationArriveeId,
      'description': description,
      'poids': poids,
      'prix': prix,
      'statut': statut,
      'date_envoi': dateEnvoi?.toIso8601String(),
      'date_livraison': dateLivraison?.toIso8601String(),
      'code_retrait': codeRetrait,
      'nom_destinataire': nomDestinataire,
      'numero_destinataire': numeroDestinataire,
    };
  }
}

File: models/Colis.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');
const Client = require('./Client');
const Chauffeur = require('./Chauffeur');
const Station = require('./Station');

const Colis = sequelize.define('Colis', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true,
  },
  client_id: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: Client,
      key: 'utilisateur_id',
    },
  },
  chauffeur_id: {
    type: DataTypes.INTEGER,
    allowNull: true,
    references: {
      model: Chauffeur,
      key: 'utilisateur_id',
    },
  },
  station_depart_id: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: Station,
      key: 'id',
    },
  },
  station_arrivee_id: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: Station,
      key: 'id',
    },
  },
  description: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  poids: {
    type: DataTypes.FLOAT,
    allowNull: false,
  },
  prix: {
    type: DataTypes.FLOAT,
    allowNull: true,
  },
  statut: {
    type: DataTypes.ENUM('en_attente', 'pris_en_charge', 'en_livraison', 'livré', 'déposé_station'),
    defaultValue: 'en_attente',
  },
  date_envoi: {
    type: DataTypes.DATE,
    allowNull: true,
  },
  date_livraison: {
    type: DataTypes.DATE,
    allowNull: true,
  },
  code_retrait: {
    type: DataTypes.STRING,
    allowNull: true, // Généré automatiquement, demandé au destinataire
  },
  nom_destinataire: {
  type: DataTypes.STRING,
  allowNull: true,
},
numero_destinataire: {
  type: DataTypes.STRING,
  allowNull: true,
  validate: {
    is: [/^\+?\d{8,15}$/],
  }
},

}, {
  tableName: 'colis',
  timestamps: false,
});
// 🔐 Génération automatique du code de retrait
Colis.beforeCreate((colis) => {
  const code = Math.floor(100000 + Math.random() * 900000).toString();
  colis.code_retrait = code;
});

module.exports = Colis;
